cpu/stm32/Kconfig.models

```diff
@@ -0,0 +1,258 @@
+# Copyright (c) 2020 Inria
+#
+# This file is subject to the terms and conditions of the GNU Lesser
+# General Public License v2.1. See the file LICENSE in the top level
+# directory for more details.
+#
+
+# STM32F0
+config CPU_MODEL_STM32F030F4
+    bool
+    select CPU_LINE_STM32F030X4
+
+config CPU_MODEL_STM32F030R8
+    bool
+    select CPU_LINE_STM32F030X8
+
+config CPU_MODEL_STM32F031K6
+    bool
+    select CPU_LINE_STM32F031X6
+
+config CPU_MODEL_STM32F042K6
+    bool
+    select CPU_LINE_STM32F042X6
+
+config CPU_MODEL_STM32F051R8
+    bool
+    select CPU_LINE_STM32F051XB
+
+config CPU_MODEL_STM32F070RB
+    bool
+    select CPU_LINE_STM32F070XB
+
+config CPU_MODEL_STM32F072RB
+    bool
+    select CPU_LINE_STM32F072XB
+
+config CPU_MODEL_STM32F091RC
+    bool
+    select CPU_LINE_STM32F091XC
+
+# STM32F1
+config CPU_MODEL_STM32F103C8
+    bool
+    select CPU_LINE_STM32F103X8
+
+config CPU_MODEL_STM32F103CB
+    bool
+    select CPU_LINE_STM32F103XB
+
+config CPU_MODEL_STM32F103RB
+    bool
+    select CPU_LINE_STM32F103XB
+
+config CPU_MODEL_STM32F103RE
+    bool
+    select CPU_LINE_STM32F103XE
+
+# STM32F2
+config CPU_MODEL_STM32F207ZG
+    bool
+    select CPU_LINE_STM32F207XX
+
+# STM32F3
+config CPU_MODEL_STM32F302R8
+    bool
+    select CPU_LINE_STM32F302X8
+
+config CPU_MODEL_STM32F303K8
```

Comment:
This is going to be messy when people start adding individual part numbers piecemeal.
I'd split that into `Kconfig.stm32f2`, `Kconfig.stm32f3` etc.

For sam0 I used a script to generate Kconfig entries for all parts where there was a vendor file #14782

This also makes working with external boards easier.

Reply:
For CPU models that is not possible since there are too many combination possible and there's no know list of all available models, at least not in the files available in the cmsis directory.
For CPU lines that's also not easily possible since some lines are providing features and others not.

But I agree that being able to generate/compute these lists automatically would be very useful.

Reply:
Also for external boards, it might still possible to just add
```kconfig
config CPU_MODEL_<complete model name>
    bool
    select CPU_LINE_<line>
```
in the board Kconfig file

Reply:
Since those are only used to select the appropriate header, I think it would be best to just name the options the same as the header then. So `CPU_MODEL_STM32F423XX`, `CPU_MODEL_STM32F401XC`, etc.

Reply:
Wait, that's already `CPU_LINE` - so what is `CPU_MODEL` used for then?

Reply:
> I think it would be best to just name the options the same as the header then

This is not so clear. The XX letters contains information about the pin count numbers and the ROM/RAM size. This information is used by the build system.

Reply:
> I'd split that into Kconfig.stm32f2, Kconfig.stm32f3 etc.

This should be possible anyway, right?

Reply:
But I agree that in Kconfig we could define the CPU line at board level. The problem I see is when features will be modeled in Kconfig, we'll still have to keep the CPU_MODEL definition in make, to determine the RAM/ROM sizes.

Reply:
I pushed a commit for grouping the definitions per families, except for the features.

Reply:
> there's no known list of all available models

In fact, all products of a given family are listed on the ST website where one can even download an Excel file with all references. So it's possible to extract a CSV and generate the complete list of models in a given family. I'll try to find some time to write a Python extractor/generator script, will be helpful and not be too difficult.